pkg/apiserver/resthandler.go

```diff
@@ -446,6 +417,93 @@ func PatchResource(r rest.Patcher, scope RequestScope, typer runtime.ObjectTyper
 
 		write(http.StatusOK, scope.APIVersion, scope.Codec, result, w, req.Request)
 	}
+
+}
+
+// patchResource divides PatchResource for easier unit testing
+func patchResource(ctx api.Context, timeout time.Duration, versionedObj runtime.Object, r rest.Patcher, name string, patchType api.PatchType, patchJS []byte, namer ScopeNamer, codec runtime.Codec) (runtime.Object, error) {
+	namespace := api.NamespaceValue(ctx)
+
+	original, err := r.Get(ctx, name)
+	if err != nil {
+		return nil, err
+	}
+
+	originalObjJS, err := codec.Encode(original)
+	if err != nil {
+		return nil, err
+	}
+	originalPatchedObjJS, err := getPatchedJS(patchType, originalObjJS, patchJS, versionedObj)
+	if err != nil {
+		return nil, err
+	}
+
+	objToUpdate := r.New()
+	if err := codec.DecodeInto(originalPatchedObjJS, objToUpdate); err != nil {
+		return nil, err
+	}
+	if err := checkName(objToUpdate, name, namespace, namer); err != nil {
+		return nil, err
+	}
+
+	return finishRequest(timeout, func() (runtime.Object, error) {
+		// update should never create as previous get would fail
+		updateObject, _, updateErr := r.Update(ctx, objToUpdate)
+		if errors.IsConflict(updateErr) {
+
+			// on a conflict,
+			// 1. build a strategic merge patch from originalJS and the patchedJS
```

Comment:
gotcha-- a parenthetical in the comment might be in order.

On Thu, Oct 1, 2015 at 5:05 PM, David Eads <notifications@github.com> wrote:

> In pkg/apiserver/resthandler.go
> <https://github.com/kubernetes/kubernetes/pull/14758#discussion_r40981279>
> :
>
> > +
> > +	objToUpdate := r.New()
> > +	if err := codec.DecodeInto(originalPatchedObjJS, objToUpdate); err != nil {
> > +		return nil, err
> > +	}
> > +	if err := checkName(objToUpdate, name, namespace, namer); err != nil {
> > +		return nil, err
> > +	}
> > +
> > +	return finishRequest(timeout, func() (runtime.Object, error) {
> > +		// update should never create as previous get would fail
> > +		updateObject, _, updateErr := r.Update(ctx, objToUpdate)
> > +		if errors.IsConflict(updateErr) {
> > +
> > +			// on a conflict,
> > +			// 1. build a strategic merge patch from originalJS and the patchedJS
>
> this is what the user passed in to begin with, no? why do we need to build
> it?
>
> They can submit a patch of a different type. I think StrategicMergePatch
> is more expressive, so I build a new patch of that type.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14758/files#r40981279>.
>